src/librustc_typeck/impl_wf_check/min_specialization.rs

```diff
@@ -0,0 +1,315 @@
+        );
+
+        let tcx = infcx.tcx;
+
+        let parent_substs = unconstrained_parent_impl_substs(tcx, impl2_def_id, impl2_substs);
+
+        check_static_lifetimes(tcx, &parent_substs, span);
+        check_duplicate_params(tcx, impl1_substs, &parent_substs, span);
+
+        check_predicates(tcx, impl1_def_id, impl1_substs, impl2_node, impl2_substs, span);
+    }
+}
+
+// Get substs for the two impls so that they have the same `TraitRef`.
+fn get_impl_substs<'tcx>(
+    infcx: &InferCtxt<'_, 'tcx>,
+    impl1_def_id: DefId,
+    impl2_node: Node,
+    span: Span,
+) -> Option<(SubstsRef<'tcx>, SubstsRef<'tcx>)> {
+    let tcx = infcx.tcx;
+    let param_env = tcx.param_env(impl1_def_id);
+
+    let impl1_substs = InternalSubsts::identity_for_item(tcx, impl1_def_id);
+    let impl2_substs = translate_substs(infcx, param_env, impl1_def_id, impl1_substs, impl2_node);
+
+    // Conservatively use an empty `ParamEnv`.
+    let outlives_env = OutlivesEnvironment::new(ty::ParamEnv::empty());
+    infcx.resolve_regions_and_report_errors(
+        impl1_def_id,
+        &ScopeTree::default(),
+        &outlives_env,
+        SuppressRegionErrors::default(),
+    );
+    let impl2_substs = match infcx.fully_resolve(&impl2_substs) {
+        Ok(s) => s,
+        Err(_) => {
+            tcx.sess.struct_span_err(span, "could not resolve substs on overridden impl").emit();
+            return None;
+        }
+    };
+    Some((impl1_substs, impl2_substs))
+}
+
+/// Returns a list of all of the unconstrained subst of the given impl.
+fn unconstrained_parent_impl_substs<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl_def_id: DefId,
+    impl_substs: SubstsRef<'tcx>,
+) -> Vec<GenericArg<'tcx>> {
+    let mut seen_params = FxHashSet::default();
+    let mut constrained_params = FxHashSet::default();
+
+    let impl2_identity_substs = InternalSubsts::identity_for_item(tcx, impl_def_id);
+    let impl2_generic_predicates = tcx.predicates_of(impl_def_id);
+    let impl2_identity_predicates =
+        impl2_generic_predicates.instantiate(tcx, impl2_identity_substs);
+
+    for predicate in impl2_identity_predicates.predicates.iter() {
+        if let ty::Predicate::Projection(proj) = predicate {
+            let projection_ty = proj.skip_binder().projection_ty;
+            let projected_ty = proj.skip_binder().ty;
+            seen_params.extend(cgp::parameters_for(&projection_ty, true));
+            for param in cgp::parameters_for(&projected_ty, false) {
+                if !seen_params.contains(&param) {
+                    constrained_params.insert(param.0);
+                }
+            }
+            seen_params.extend(cgp::parameters_for(&projected_ty, true));
+        }
+    }
+
+    impl_substs
+        .iter()
+        .enumerate()
+        .filter(|(idx, _)| !constrained_params.contains(&(*idx as u32)))
+        .map(|(_, arg)| *arg)
+        .collect()
+}
+
+fn check_duplicate_params<'tcx>(
```

Comment:
Comment please -- it'd be great to document how this function fits into the overall algorithm. I find myself a bit confused about it. Also, is there a reason to use `&Vec<..>` and not `&[..]`? It doesn't matter of course, I just found myself distracted (I assumed at first it was `&mut Vec<...>`...), and it seems like a slice would be more obvious.

Reply:
I mean I see that it's role is to check the duplicate parameter case, but I can't quite relate `impl1_substs` and `parent_substs` etc. It'd be awesome to have a Rust example that shows what each thing is.

Reply:
It needs to be `Vec` so that it's `TypeFoldable`.